### PR TITLE
Add support for variables in stack frames

### PR DESCRIPTION
**Pull request by alan-knight**
Sorry for the size, this has several different things in it
 - add variables in dartFramesFor (the important part)
 - add a single test for that
 - operate better without a connection, so we can write unit tests more easily
 - provide fake implementations of WipConnection, AppInspector, and Runtime for tests
 - make AppInspector extend Domain and use checkIsolate
 - add a library for WebKit mirror objects not yet in wip_inspection_protocol (Property)
 -  add an example with more complicated scopes than the hello world example. Mostly used for manual testing so far, though I did extract hard-coded test data from it.

### CODE REVIEW

File: dwds/lib/src/debugger.dart [hunk 1]
```diff
@@ -282,12 +284,51 @@ class Debugger extends Domain {
...
         dartFrame.index = index++;
+        dartFrame.vars =
+            await _variablesFor(frame['scopeChain'] as List<dynamic>);
         dartFrames.add(dartFrame);
       }
     }
     return dartFrames;
   }
 
+  /// The variables visible in a frame in Dart protocol BoundVariable form.
```
Comment (grouma) [inline dwds/lib/src/debugger.dart]: nit: [BoundVariable]

File: dwds/lib/src/debugger.dart [hunk 2]
```diff
@@ -282,12 +284,51 @@ class Debugger extends Domain {
...
+  Future<List<BoundVariable>> _variablesFor(List<dynamic> scopeChain) async {
+    // TODO: Much better logic for which frames to use. This is probably just
+    // the dynamically visible variables, so we should omit library scope.
+    return [
+      for (var scope in scopeChain.take(2)) ...await _boundVariables(scope)
```
Comment (grouma) [inline dwds/lib/src/debugger.dart]: Modern Dart! :+1:

File: dwds/lib/src/debugger.dart [hunk 3]
```diff
@@ -282,12 +284,51 @@ class Debugger extends Domain {
...
+    var refs = properties
+        .map<Future<BoundVariable>>((property) async => BoundVariable()
+          ..name = property.name
+          ..value = await inspector.instanceRefFor(property.value));
+    return Future.wait(refs);
+  }
+
+  /// Calls the Chrome Runtime.getProperties API for the object
+  /// with [id].
+  Future<List<Property>> getProperties(String id) async {
```
Comment (grouma) [inline dwds/lib/src/debugger.dart]: Can this be private?

File: dwds/lib/src/domain.dart [hunk 4]
```diff
@@ -16,6 +16,10 @@ abstract class Domain {
 
   Domain(this._appInspectorProvider);
 
+  /// A constructor for the AppInspector to call which doesn't set
+  /// [_appInspectorProvider] as it's not used by the AppInspector.
+  Domain.forInspector();
```
Comment (grouma) [inline dwds/lib/src/domain.dart]: I don't follow this. Why is this required?

File: dwds/lib/src/inspector.dart [hunk 5]
```diff
@@ -157,16 +173,26 @@ function($argsString) {
       case 'boolean':
         return _primitiveInstance(InstanceKind.kBool, remoteObject);
       case 'object':
+        // TODO: Actual toString()
+        var toString = 'Placeholder for toString() result';
+        var truncated = toString.substring(0, math.min(100, toString.length));
```
Comment (grouma) [inline dwds/lib/src/inspector.dart]: Curious where this 100 value is coming from.

File: dwds/lib/src/objects.dart [hunk 6]
```diff
@@ -0,0 +1,31 @@
+// Copyright (c) 2019, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.import 'dart:async';
+
+/// A library for WebKit mirror objects and support code. These probably should
+/// get migrated into webkit_inspection_protocol over time.
+
+import 'package:webkit_inspection_protocol/webkit_inspection_protocol.dart';
```
Comment (grouma) [inline dwds/lib/src/objects.dart]: Can we not depend on this package for this abstraction? This will become problematic as we make use of the Chrome Debug Extension.

File: dwds/lib/src/debugger.dart (hunk 1 above)
Comment (alan-knight) [reply dwds/lib/src/debugger.dart]: Done.

File: dwds/lib/src/debugger.dart (hunk 3 above)
Comment (alan-knight) [reply dwds/lib/src/debugger.dart]: Done.

File: dwds/lib/src/domain.dart (hunk 4 above)
Comment (alan-knight) [reply dwds/lib/src/domain.dart]: Domain takes an AppInspectorProvider. The AppInspectorProvider has to return the instance. When we're in the constructor for AppInspector we can't reference this, so we can't call super with an AppInspectorProvider. The only workaround I found was to give it a constructor that didn't need that.

File: dwds/lib/src/inspector.dart (hunk 5 above)
Comment (alan-knight) [reply dwds/lib/src/inspector.dart]: It's completely arbitrary. Added a TODO to make that consistent with what the VM does.

File: dwds/lib/src/objects.dart (hunk 6 above)
Comment (alan-knight) [reply dwds/lib/src/objects.dart]: It needs that now because it can return RemoteObject from that package for its values. My thought was more to fold this into that package than to re-implement what it does.

Why is it problematic if we use the extension? Is the extension going to use an incompatible protocol?

Comment (grouma) [reply dwds/lib/src/objects.dart]: Discussed offline. We'll likely reuse the abstractions provided by webkit_inspection_protocol but change out the actual communication layer.
